src/Microsoft.Azure.ServiceBus/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -45,11 +45,11 @@ void ScheduleCleanup()
                 }
 
                 this.cleanupScheduled = true;
-                Task.Run(() => this.CollectExpiredEntries());
+                Task.Run(async () => await this.CollectExpiredEntries());
             }
```

Comment:
This wont completely help. Since every `AddOrUpdate()` call will end up calling `ScheduleCleanup`, and in a high throughput scenario, the moment one thread has finished setting `this.cleanupScheduled=false`, another thread might start working on `CollectExpiredEntries`. We didnt really end up waiting for the timespan.

I'd recommend moving 
```
if (this.dictionary.Count <= 0)
            {
                return;
            }
```
inside `ScheduleCleanup` just before we set `this.cleanupScheduled = true`.

And move `Task.Delay()` as the first line in `CollectExpiredEntries`

Reply:
Makes sense, done